src/wallet/mod.rs

```diff
@@ -972,38 +957,53 @@ where
             (utxo, weight)
         };
 
-        match utxo {
-            // with manual coin selection we always want to spend all the selected utxos, no matter
-            // what (even if they are marked as unspendable)
-            Some(raw_utxos) => {
-                let full_utxos = raw_utxos
-                    .iter()
-                    .map(|u| {
-                        Ok(add_weight(
-                            self.database
-                                .borrow()
-                                .get_utxo(&u)?
-                                .ok_or(Error::UnknownUTXO)?,
-                        ))
-                    })
-                    .collect::<Result<Vec<_>, Error>>()?;
+        let utxos = self.list_unspent()?.into_iter().map(add_weight).collect();
 
-                Ok((full_utxos, true))
-            }
-            // otherwise limit ourselves to the spendable utxos for the selected policy, and the `send_all` setting
-            None => {
-                let utxos = self.list_unspent()?.into_iter();
-                let utxos = change_policy.filter_utxos(utxos).into_iter();
-
-                Ok((
-                    utxos
-                        .filter(|u| !unspendable_set.contains(&u.outpoint))
-                        .map(add_weight)
-                        .collect(),
-                    send_all,
-                ))
-            }
+        Ok(utxos)
+    }
+
+    /// Given the options returns the list of utxos that must be used to form the
+    /// transaction and any further that may be used if needed.
+    fn get_must_may_use_utxos(
+        &self,
+        change_policy: tx_builder::ChangeSpendPolicy,
+        unspendable: &HashSet<OutPoint>,
+        manually_selected: &[OutPoint],
+        send_all: bool,
+        manual_only: bool,
+    ) -> Result<(Vec<(UTXO, usize)>, Vec<(UTXO, usize)>), Error> {
```

Comment:
Did you consider introducing an enum like 
```
enum UseUTXO {
  May(UTXO, usize),
  Must(UTXO, usize),
}
```
so that the Result would be just Vec<UseUTXO> ?

Reply:
`coin_select` actually uses the two distinct vectors. It sorts the `may_use` in order of size before selecting them. `must_use` are not sorted. So putting them altogether in the same vector would complicate that a bit so I think it's best to leave it as it is now at least for this PR.